design/one-pager-package-image-config.md

```diff
@@ -0,0 +1,377 @@
+as experimental.
+
+In Crossplane itself, we had [a previous proposal] aiming to solve signature
+verification for packages but mostly focusing on the implementation details,
+primarily for verification using a shared public key. The follow-up
+[implementation PR] extended the solution also to support keyless verification
+by mostly following the same solution / similar API as Flux by extending the
+Package installation APIs for verification settings, leaving the dependencies
+out of the scope. This proposal aims to provide a more comprehensive solution by
+introducing an API that can be extended to support various verification settings
+and also covering the ones coming as a dependency.
+
+## Goals
+
+- Allow configuration of credentials for package images and their dependencies
+  at runtime.
+- Enable users to set policies for verifying image signatures.
+- Define an image-centric configuration API that centralizes settings for
+  package images.
+
+## Proposal
+
+This document proposes a new API, `ImageConfig`, under the `pkg.crossplane.io`
+API group, that allows users to configure settings for package images. The API
+enables users to define rules for matching images and configuring how to
+interact with the registries hosting the images, including authentication and
+TLS settings. It also provides a way to define policies for verifying image
+signatures.
+
+Considering the API is solving some existing problems, e.g. configuring pull
+secrets for dependencies, and not introducing a behavioral change unless the
+users opt in by creating an `ImageConfig` object, we believe this API should be
+introduced as a beta API without going through the feature lifecycle stages.
+
+### API
+
+```yaml
+apiVersion: pkg.crossplane.io/v1beta1
+kind: ImageConfig
+metadata:
+  name: acme-packages
+spec:
+  matchImages:
+    - prefix: registry1.com/acme-co/configuration-foo
+    - prefix: registry1.com/acme-co/configuration-bar
+    - prefix: registry1.com/acme-co/function-baz
+  registry:
+    authentication:
+      pullSecretRef:
+        name: acme-registry-credentials
+    tls:
+      enabled: true
+      insecure: false
+      caBundleConfigMapRef:
+        name: registry1-ca
+        key: ca.crt
+  verification:
+    provider: cosign
+    cosign:
+      authorities:
+        - name: verify acme builds
+          keyless:
+            url: https://fulcio.sigstore.dev
+            identities:
+              - issuer: https://token.actions.githubusercontent.com
+                subjectRegExp: https://github.com/acme-co/crossplane-packages/*
+          attestations:
+            - name: verify attestations
+              predicateType: spdxjson
+```
+
+The `ImageConfig` API `spec` has the following fields:
+
+- `matchImages`: A list of image prefixes to match. The package manager will
+  apply the settings defined in the `ImageConfig` object to the images that
+  match the prefixes. The longest prefix match will be selected in case of
```

Comment:
What if down the line we introduced regex matching? What would be the meaning of longest match then?

Reply:
I presume we'd be explicit add a `type: Regex` which could have different semantics from `type: Prefix` when matching. So we could decide longest match wasn't relevant in that case.

Reply:
But it could be mixed with prefix matchers either in the same ImageConfig or across multiple ones.

Reply:
A list is fine, avoids the need for complicated patterns with ORs, all the elements in the list are effectively OR-ed and if at least one of them matches an image the ImageConfig matches. If we allow multiple ImageConfigs to match and pick one according to some rule, if the rule is based on the match we'll have to pick a match from the list and compare its score against the one from the other ImageConfig. Using some other parameter to compare matching ImageConfigs would make it better.

To avoid useless complexity, I'd say we just error out if multiple matching ImageCatalogs are found, this way we don't need to think of longest matches or any other scoring mechanism. 

And to have a simpler ux for users  and to allow us to add other matching mechanisms we keep the `match[*].prefix`, with the added explicit type, as suggested by @negz.

Reply:
> I don't want to push us into a corner with a strict API, but, I may even prefer to make this explicit by changing the `matchImages[*].prefix` to `matchImagesWithPrefix[]`. 

Unless we want to have something like below at some point:

```yaml
matchImages:
  - prefix: registry1.com/acme-co/configuration-foo
    constraint: "=< v1"
```

Reply:
I am still leaning towards relying on the longest match, maybe until we realize we need more advanced matching mechanisms like regexes/globes. I believe I am feeling good about longest match because registries-conf is using the same mechanism and widely adopted: https://www.mankier.com/5/containers-registries.conf#Description-Choosing_a_%5B%5Bregistry%5D%5D_TOML_table

As long as we communicate which ImageConfig is picked up, or in case of signature verification, whether the signature verified or skipped, I don't think multiple matches is that much of a problem and people can fix it if needed.

Reply:
Resolving in favor of https://github.com/crossplane/crossplane/pull/5993#issuecomment-2406859296